Updating kdbush and geokdbush types on DefinitelyTyped

Hi @mourner. Have taken your advice to use kdbush in Turf to improve performance. Results look very promising so far. Thank you!

Did come across a typing issue though between kdbush and geokdbush. Namely around KDBush no longer being a named export or a generic.

Wanted to check in to go through steps to tidy this up. I think I need to:

1. create a DT PR to snapshot the kdbush DT types as v3 - like has been done [here for v1](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/kdbush/v1). Process described [here](https://github.com/DefinitelyTyped/DefinitelyTyped?tab=readme-ov-file#if-a-library-is-updated-to-a-new-major-version-with-breaking-changes-how-should-i-update-its-type-declaration-package). Think that should keep everyone still on v3 happy.
2. once the above is merged, request that kdbush no longer requires its types on DT as it packages it's own as of v4. This step is described at [Removing a package](https://github.com/DefinitelyTyped/DefinitelyTyped?tab=readme-ov-file#removing-a-package). Believe the `asOfVersion` should be 4.0.0. @types/kdbush are currently at 3.0.5 so that should suit the criteria.
3. once the above takes effect, create a second DT PR to update the geokdbush types to import KDBush as the default export, and define index as KDBush without the generic.

Does that sit well with you? Anything I've missed?

I'm running into a similar issue after updating the dependencies of my project:

```
> tsc && vite build

node_modules/@types/geokdbush/index.d.ts:6:10 - error TS2614: Module '"kdbush"' has no exported member 'KDBush'. Did you mean to use 'import KDBush from "kdbush"' instead?

6 import { KDBush } from 'kdbush';
           ~~~~~~

Found 1 error in node_modules/@types/geokdbush/index.d.ts:6
```

Any idea how to fix this?

--- a/.gitignore
+++ b/.gitignore
@@ -1,1 +1,2 @@
 node_modules
+index.d.ts

--- a/index.js
+++ b/index.js
@@ -212,7 +212,7 @@ function haverSinDist(lng1, lat1, lng2, lat2, cosLat1) {
 }
 
 /**
- * Returns gerat circle distance between two locations in kilometers.
+ * Returns great circle distance between two locations in kilometers.
  *
  * @param {number} lng1
  * @param {number} lat1

--- a/index.js
+++ b/index.js
@@ -4,6 +4,32 @@ import TinyQueue from 'tinyqueue';
 const earthRadius = 6371;
 const rad = Math.PI / 180;
 
+/**
+ * @typedef {object} Node
+ * @prop {number} left left index in the kd-tree array
+ * @prop {number} right right index in the kd-tree array
+ * @prop {number} axis 0 for longitude and 1 for latitude axis
+ * @prop {number} dist the lower bound of children's distances to the query point
+ * @prop {number} minLng
+ * @prop {number} minLat
+ * @prop {number} maxLng
+ * @prop {number} maxLat
+ */
+
+/**
+ * @typedef {{ id: number, dist: number }} QueueItem
+ */
+
+/**
+ * Returns an array of the closest points from a given location in order of increasing distance.
+ *
+ * @param {import('kdbush').default} index kdbush index
+ * @param {number} lng Query point longitude
+ * @param {number} lat Query point latitude
+ * @param {number} [maxResults] Maximum number of points to return
+ * @param {number} [maxDistance] Maximum distance in kilometers to search within
+ * @param {(item: number) => boolean} [predicate] A function to filter the results
+ */
 export function around(index, lng, lat, maxResults = Infinity, maxDistance = Infinity, predicate) {
     let maxHaverSinDist = 1;
     const result = [];
@@ -11,16 +37,22 @@ export function around(index, lng, lat, maxResults = Infinity, maxDistance = Inf
     if (maxResults === undefined) maxResults = Infinity;
     if (maxDistance !== undefined) maxHaverSinDist = haverSin(maxDistance / earthRadius);
 
-    // a distance-sorted priority queue that will contain both points and kd-tree nodes
+    /**
+     * a distance-sorted priority queue that will contain both points and kd-tree nodes
+     * @type {TinyQueue<QueueItem|Node>}
+     */
     const q = new TinyQueue([], compareDist);
 
-    // an object that represents the top kd-tree node (the whole Earth)
+    /**
+     * an object that represents the top kd-tree node (the whole Earth)
+     * @type {Node|undefined}
+     */
     let node = {
-        left: 0, // left index in the kd-tree array
-        right: index.ids.length - 1, // right index
-        axis: 0, // 0 for longitude axis and 1 for latitude axis
-        dist: 0, // will hold the lower bound of children's distances to the query point
-        minLng: -180, // bounding box of the node
+        left: 0,
+        right: index.ids.length - 1,
+        axis: 0,
+        dist: 0,
+        minLng: -180,
         minLat: -90,
         maxLng: 180,
         maxLat: 90
@@ -92,21 +124,27 @@ export function around(index, lng, lat, maxResults = Infinity, maxDistance = Inf
         // fetch closest points from the queue; they're guaranteed to be closer
         // than all remaining points (both individual and those in kd-tree nodes),
         // since each node's distance is a lower bound of distances to its children
-        while (q.length && q.peek().id != null) {
-            const candidate = q.pop();
+        let candidate;
+        while ((candidate = q.pop()) && 'id' in candidate) {
             if (candidate.dist > maxHaverSinDist) return result;
             result.push(candidate.id);
             if (result.length === maxResults) return result;
         }
 
         // the next closest kd-tree node
-        node = q.pop();
+        node = candidate;
     }
 
     return result;
 }
 
-// lower bound for distance from a location to points inside a bounding box
+/**
+ * lower bound for distance from a location to points inside a bounding box
+ * @param {number} lng
+ * @param {number} lat
+ * @param {number} cosLat
+ * @param {{ minLng: number, maxLng: number, minLat: number, maxLat: number }} node
+ */
 function boxDist(lng, lat, cosLat, node) {
     const minLng = node.minLng;
     const maxLng = node.maxLng;
@@ -136,29 +174,61 @@ function boxDist(lng, lat, cosLat, node) {
     );
 }
 
+/**
+ * @param {QueueItem|Node} a
+ * @param {QueueItem|Node} b
+ */
 function compareDist(a, b) {
     return a.dist - b.dist;
 }
 
+/** @param {number} theta */
 function haverSin(theta) {
     const s = Math.sin(theta / 2);
     return s * s;
 }
 
+/**
+ * @param {number} haverSinDLng
+ * @param {number} cosLat1
+ * @param {number} lat1
+ * @param {number} lat2
+ */
 function haverSinDistPartial(haverSinDLng, cosLat1, lat1, lat2) {
     return cosLat1 * Math.cos(lat2 * rad) * haverSinDLng + haverSin((lat1 - lat2) * rad);
 }
 
+
+/**
+ * @param {number} lng1
+ * @param {number} lat1
+ * @param {number} lng2
+ * @param {number} lat2
+ * @param {number} cosLat1
+ */
 function haverSinDist(lng1, lat1, lng2, lat2, cosLat1) {
     const haverSinDLng = haverSin((lng1 - lng2) * rad);
     return haverSinDistPartial(haverSinDLng, cosLat1, lat1, lat2);
 }
 
+/**
+ * Returns gerat circle distance between two locations in kilometers.
+ *
+ * @param {number} lng1
+ * @param {number} lat1
+ * @param {number} lng2
+ * @param {number} lat2
+ */
 export function distance(lng1, lat1, lng2, lat2) {
     const h = haverSinDist(lng1, lat1, lng2, lat2, Math.cos(lat1 * rad));
     return 2 * earthRadius * Math.asin(Math.sqrt(h));
 }
 
+
+/**
+ * @param {number} lat
+ * @param {number} haverSinDLng
+ */
 function vertexLat(lat, haverSinDLng) {
     const cosDLng = 1 - 2 * haverSinDLng;
     if (cosDLng <= 0) return lat > 0 ? 90 : -90;

--- a/package.json
+++ b/package.json
@@ -11,6 +11,9 @@
   "dependencies": {
     "tinyqueue": "^2.0.3"
   },
+  "files": [
+    "index.d.ts"
+  ],
   "repository": {
     "type": "git",
     "url": "https://github.com/mourner/geokdbush.git"

--- a/package.json
+++ b/package.json
@@ -3,12 +3,14 @@
   "version": "2.0.1",
   "type": "module",
   "main": "index.js",
+  "exports": "./index.js",
+  "types": "index.d.ts",
+  "sideEffects": false,
   "author": "Vladimir Agafonkin <agafonkin@gmail.com>",
   "license": "ISC",
   "dependencies": {
     "tinyqueue": "^2.0.3"
   },
-  "files": [],
   "repository": {
     "type": "git",
     "url": "https://github.com/mourner/geokdbush.git"
@@ -19,11 +21,12 @@
     "eslint-config-mourner": "^4.0.1",
     "kdbush": "^4.0.2",
     "sphere-knn": "^1.4.0",
+    "typescript": "^6.0.2",
     "vptree": "^1.0.0"
   },
   "scripts": {
     "pretest": "eslint index.js test.js bench",
-    "test": "node test.js",
+    "test": "tsc && node test.js",
     "bench": "cd bench && node bench-geokdbush && node bench-sphere-knn && node bench-naive && node bench-vptree"
   }
 }

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,0 +1,13 @@
+{
+    "compilerOptions": {
+        "allowJs": true,
+        "checkJs": true,
+        "strict": true,
+        "emitDeclarationOnly": true,
+        "declaration": true,
+        "target": "es2017"
+    },
+    "files": [
+        "index.js"
+    ]
+}

